modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/components/PageRenderer/Actions.es.js

```diff
@@ -0,0 +1,187 @@
+			});
+		}
+	}, [hoveredField, dispatch]);
+
+	return (
+		<ActionsContext.Provider
+			value={{activeField, hoveredField, setActiveField, setHoveredField}}
+		>
+			{children}
+		</ActionsContext.Provider>
+	);
+};
+
+export const useActions = () => {
+	return useContext(ActionsContext);
+};
+
+ActionsContext.displayName = 'ActionsContext';
+
+const ACTIONS_CONTAINER_OFFSET = [-2, 1];
+
+export const ActionsControls = forwardRef(
+	({activePage, children, columnRef, editable, field}, actionsRef) => {
+		const {
+			activeField,
+			hoveredField,
+			setActiveField,
+			setHoveredField,
+		} = useActions();
+
+		const dispatch = useForm();
+
+		const contentRect = useResizeObserver(columnRef);
+
+		useLayoutEffect(() => {
+			if (actionsRef.current && columnRef.current) {
+				domAlign(actionsRef.current, columnRef.current, {
+					offset: ACTIONS_CONTAINER_OFFSET,
+					points: ['bl', 'tl'],
+				});
+			}
+		}, [actionsRef, activeField, columnRef, contentRect, hoveredField]);
+
+		const handleFieldInteractions = (event) => {
+			event.stopPropagation();
+
+			if (actionsRef.current?.contains(event.target) && !editable) {
+				return;
+			}
+
+			switch (event.type) {
+				case 'click':
+					dispatch({
+						payload: {activePage, fieldName: field.fieldName},
+						type: EVENT_TYPES.FIELD_CLICKED,
+					});
+
+					setActiveField(field.fieldName);
+
+					break;
+
+				case 'mouseover':
+					setHoveredField(field.fieldName);
+
+					break;
+
+				case 'mouseleave':
+					setHoveredField('');
+
+					break;
+
+				default:
+					break;
+			}
+		};
+
+		return React.cloneElement(children, {
+			onClick: handleFieldInteractions,
+			onMouseLeave: handleFieldInteractions,
+			onMouseOver: handleFieldInteractions,
+		});
+	}
+);
+
+export const Actions = forwardRef(
+	({activePage, children, expanded, field, isFieldSet}, actionsRef) => {
+		const dispatch = useForm();
+
+		const DROPDOWN_ACTIONS = [
+			{
+				label: Liferay.Language.get('duplicate'),
+				onClick: () =>
+					dispatch({
+						payload: {activePage, fieldName: field.fieldName},
+						type: EVENT_TYPES.FIELD_DUPLICATED,
+					}),
+			},
+			{
+				label: Liferay.Language.get('delete'),
+				onClick: () =>
+					dispatch({
+						payload: {activePage, fieldName: field.fieldName},
+						type: EVENT_TYPES.FIELD_DELETED,
+					}),
+			},
+		];
+
+		return (
+			<>
+				{expanded && (
+					<div
+						className={classNames('ddm-field-actions-container', {
+							'ddm-fieldset': isFieldSet,
+						})}
+						ref={actionsRef}
+					>
+						<span className="actions-label">{field.label}</span>
+
+						<ClayDropDownWithItems
+							className="dropdown-action"
+							items={DROPDOWN_ACTIONS}
+							trigger={
+								<ClayButtonWithIcon
+									aria-label={Liferay.Language.get('actions')}
+									data-title={Liferay.Language.get('actions')}
+									displayType="unstyled"
+									symbol="ellipsis-v"
+								/>
+							}
+						/>
+					</div>
+				)}
+
+				{children}
```

Comment:
We can safely remove this `children` here we don't need it anymore.

Reply:
Thanks for pointing up this!

Reply:
Done

Reply:
✅